### PR TITLE
Upgrade to mongoc-1.3.0-beta0 and libbson-1.3.0-beta0

### PR DESCRIPTION
**Pull request by bjori**
There is a known memory leak in mongoc, see https://jira.mongodb.org/browse/CDRIVER-1017

### CODE REVIEW

File: .travis.scripts/compile.sh [hunk 1]
```diff
@@ -20,7 +20,7 @@ build_lcov() {
 build_lcov
 
 phpize
-./configure --enable-developer-flags --enable-coverage
+./configure --enable-developer-flags --enable-coverage CFLAGS="-Wno-type-limits"
```
Comment (bjori) [inline .travis.scripts/compile.sh]: https://jira.mongodb.org/browse/CDRIVER-1016

File: php_phongo.c [hunk 2]
```diff
@@ -916,12 +918,7 @@ bool phongo_stream_socket_check_closed(mongoc_stream_t *stream) /* {{{ */
 	php_phongo_stream_socket *base_stream = (php_phongo_stream_socket *)stream;
 	TSRMLS_FETCH_FROM_CTX(base_stream->tsrm_ls);
 
-	return PHP_STREAM_OPTION_RETURN_OK == php_stream_set_option(base_stream->stream, PHP_STREAM_OPTION_CHECK_LIVENESS, 0, NULL);
-} /* }}} */
-
-mongoc_stream_t* phongo_stream_get_base_stream(mongoc_stream_t *stream) /* {{{ */
```
Comment (bjori) [inline php_phongo.c]: This function isn't needed and was a fool's errand to begin with based on misunderstanding, sorry!

File: php_phongo.c [hunk 3]
```diff
@@ -1356,10 +1352,8 @@ void php_phongo_cursor_to_zval(zval *retval, php_phongo_cursor_t *cursor) /* {{{
 		_ADD_BOOL(zcursor, is_command);
 		_ADD_BOOL(zcursor, sent);
 		_ADD_BOOL(zcursor, done);
-		_ADD_BOOL(zcursor, failed);
```
Comment (bjori) [inline php_phongo.c]: I don't think this was ever actually used to signal failure, so it was removed

File: php_phongo.c [hunk 4]
```diff
@@ -1356,10 +1352,8 @@ void php_phongo_cursor_to_zval(zval *retval, php_phongo_cursor_t *cursor) /* {{{
...
 		_ADD_BOOL(zcursor, end_of_event);
 		_ADD_BOOL(zcursor, in_exhaust);
-		_ADD_BOOL(zcursor, redir_primary);
```
Comment (bjori) [inline php_phongo.c]: We don't redirect commands anymore, as per spec.
See https://jira.mongodb.org/browse/CDRIVER-807

File: src/BSON/Javascript.c [hunk 5]
```diff
@@ -73,6 +73,7 @@ PHP_METHOD(Javascript, __construct)
 	}
 
 	php_phongo_new_javascript_from_javascript_and_scope(0, getThis(), javascript, javascript_len, &scope TSRMLS_CC);
+	bson_destroy(&scope);
```
Comment (bjori) [inline src/BSON/Javascript.c]: This was pre-existing memory leak.

File: src/MongoDB/BulkWrite.c [hunk 6]
```diff
@@ -315,7 +315,7 @@ HashTable *php_phongo_bulkwrite_get_debug_info(zval *object, int *is_temp TSRMLS
 		add_assoc_null_ex(&retval, ZEND_STRS("collection"));
 	}
 
-	add_assoc_bool_ex(&retval, ZEND_STRS("ordered"), intern->bulk->ordered);
+	add_assoc_bool_ex(&retval, ZEND_STRS("ordered"), intern->bulk->flags.ordered);
```
Comment (bjori) [inline src/MongoDB/BulkWrite.c]: flags also contains bypassDocumentValidation.
I imagine you'd want to add that here too as part of https://jira.mongodb.org/browse/PHPC-443

File: src/MongoDB/Manager.c [hunk 7]
```diff
@@ -267,8 +268,12 @@ PHP_METHOD(Manager, selectServer)
 	}
 
 	readPreference = phongo_read_preference_from_zval(zreadPreference TSRMLS_CC);
-	server_id = mongoc_cluster_preselect(&intern->client->cluster, MONGOC_OPCODE_QUERY, readPreference, NULL);
-	phongo_server_init(return_value, intern->client, server_id TSRMLS_CC);
+	selected_server = mongoc_topology_select(intern->client->topology, MONGOC_SS_READ, readPreference, MONGOC_SS_DEFAULT_LOCAL_THRESHOLD_MS, &error);
+	if (selected_server) {
+		phongo_server_init(return_value, intern->client, selected_server->id TSRMLS_CC);
+	} else {
+		phongo_throw_exception(PHONGO_ERROR_RUNTIME TSRMLS_CC, "%s", error.message);
```
Comment (bjori) [inline src/MongoDB/Manager.c]: https://jira.mongodb.org/browse/PHPC-488

File: src/MongoDB/Manager.c [hunk 8]
```diff
@@ -267,8 +268,12 @@ PHP_METHOD(Manager, selectServer)
 	}
 
 	readPreference = phongo_read_preference_from_zval(zreadPreference TSRMLS_CC);
-	server_id = mongoc_cluster_preselect(&intern->client->cluster, MONGOC_OPCODE_QUERY, readPreference, NULL);
-	phongo_server_init(return_value, intern->client, server_id TSRMLS_CC);
+	selected_server = mongoc_topology_select(intern->client->topology, MONGOC_SS_READ, readPreference, MONGOC_SS_DEFAULT_LOCAL_THRESHOLD_MS, &error);
```
Comment (bjori) [inline src/MongoDB/Manager.c]: preselect has been removed as it became redundant after internal refactoring.
Now we return a mongoc_server_description_t (on success), which contains the server_id

File: src/MongoDB/Query.c [hunk 9]
```diff
@@ -75,7 +75,6 @@ PHP_METHOD(Query, __construct)
 	zval_to_bson(zfilter, PHONGO_BSON_NONE, &bfilter, NULL TSRMLS_CC);
 
 	if (zoptions) {
-		bson_init(&boptions);
```
Comment (bjori) [inline src/MongoDB/Query.c]: Pre-existing double initialization

**Comment by derickr** [comment]
I am closing this in favour of #141, as that already had some work done towards the changes in this PR.

(You might also want to rename your forked repo from mongo-php-driver-prototype to mongo-php-driver, as I got confused when pulling in the legacy commits while cherry picking your commits :-) )

**Comment by bjori** [comment]
I actually confused my self to when I couldn't find my changes on github.com/bjori/mongo-php-driver -- finally figuring it was the legacy driver haha

**Pull request closed without merging**